python-lib/design/ab_dispatcher.py

```diff
@@ -0,0 +1,106 @@
+
+        :param np.array experiment_population: array containing references to the experiment population
+        :param AttributionMethod attribution: defines how to deal with the leftover users.
+        :raises: :class:`ValueError`: Not enough users to run the experiment
+
+        :returns: Experiment dataset and two arrays containing the contacts for each group
+        :rtype: tuple
+        """
+        self.check_sample_size(experiment_population)
+        A_group, B_group = self.split_into_groups(experiment_population, attribution)
+        group_df = self.add_group_column(input_df, A_group, B_group)
+        logger.info("Dispatching experiment population: done")
+        return group_df, A_group, B_group
+
+    def check_sample_size(self, experiment_population: np.array):
+        """Check if the size of the population is large enough to run the experiment.
+
+        :raises: :class:`ValueError`: Not enough users to run the experiment
+        """
+        population_size = experiment_population.shape[0]
+
+        logger.info("Dispatching experiment population ...")
+        logger.info("Dispatching experiment population: checking the sample sizes")
+        logger.info("Sample size of the A group:" + str(self.size_A))
+        logger.info("Sample size of the B group:" + str(self.size_B))
+        logger.info("Size of the experiment population: " + str(population_size))
+
+        if self.size_A + self.size_B > population_size:
+            raise ValueError("Not enough user ids or emails to run the current AB testing.")
+
+    def split_into_groups(self, experiment_population: np.array, attribution_method: str) -> tuple:
+        """Shuffle the population and splits it into groups
+
+        :returns: Two arrays containing the contacts for each group
+        :rtype: tuple
+        """
+        logger.info("Dispatching experiment population: shuffle")
+        np.random.permutation(experiment_population)
+        if attribution_method == AttributionMethod.LEFTOVER_TO_A:
+            B_group = experiment_population[:self.size_B]
+            A_group = experiment_population[self.size_B:]
+        elif attribution_method == AttributionMethod.LEFTOVER_TO_B:
+            A_group = experiment_population[:self.size_A]
+            B_group = experiment_population[self.size_A:]
+        elif attribution_method == AttributionMethod.LEFTOVER_BLANK:
+            B_group = experiment_population[:self.size_B]
+            A_group = experiment_population[self.size_B:self.size_A+self.size_B]
+        return A_group, B_group
+
+    def add_group_column(self, input_df: pd.DataFrame, A_group: np.array, B_group: np.array) -> pd.DataFrame:
+        """Add a column to the input dataframe to indicate the group for each user
+
+        :returns: The input dataframe with a "AB_group" column
+        :rtype: pd.DataFrame
+        """
+        input_df["AB_group"] = input_df[self.reference_column].apply(lambda x: retrieve_group(x, A_group, B_group))
```

Comment:
I am trying to do such a randomly generated binary mask. I've been using np.random.binomial with the probability that I want ( For instance, 0.5 if total size is 500 and the sample sizes required for each group are respectively 250). But as it is random, I don't end up with the exact number that I want (I might get 253 instead of 250). So I could : 
-  just make up for the imbalance afterwards by taking 3 from A and adding them to B, but it's not so pretty. 
- use random.choice(index, 250) on the index to pick randomly 250 elements from the index

Reply:
done in https://github.com/dataiku/dss-plugin-ab-testing/pull/1/commits/42e6adc2eb62d94294976c37afbd072cbbf53120